### PR TITLE
Add runtime validity checks inside MaybeUninit::assume_init

### PR DESCRIPTION
**Pull request by ghost**
This is very much an experiment, I have no idea if this will work or is a good idea.

However, it (when combined with memory sanitizer), *did* find the use of uninit memory inside `http`. It introduces branches based on all fields it finds, even if those fields are of a type that's valid at any byte value. Granted, that was *also* found with miri, but this can work where miri doesn't.

This probably absolutely kills perf, so calls to assert_validity_of might need to be configured out unless a flag is given when building the stdlib.

One thing we could do is write out a maximally invalid type inside MaybeUninit::uninit, so if you do `MaybeUninit<bool>::uninit()`, it writes out the value `2`, so you don't even need memory sanitizer to detect assume_init of uninit memory, *if* there is a niche value we can write to mean "uninit". That would mean miri and memory sanitizer *can't* detect the uninit memory, which would be a problem. Still, would be neat.

---

Still need to:

- [x] Deduplicate invariants generated
- [x] Don't generate invariants for types like u8 when not running under memory sanitizer
- [ ] Clean up the unwraps
- [ ] Check out the perf impact

### CODE REVIEW

**Comment by 5225225** [comment]
This is mainly just a safety net to try and catch UB. There is already some checks similar to this in the stdlib already, see #92686. This is just a pretty extreme runtime check, I'm not proposing to make it a safe method.

One case where this *can't* detect UB is use of uninit memory. Without memory sanitizer, we just *can't* notice when someone forgot to write to a MaybeUninit<u8>.

Another case that's UB that we can't detect here (even with memory sanitizer), is use of pointer bytes as an integer. Miri can detect both of these, since it has the extra runtime tracking to know what bytes are what.

Also, we're not checking enums here, but that's not a strict limitation, it just complicates the code since you need to have conditionals (check byte `1` is in range `0..1` if byte `0` is `42`).

File: library/core/src/intrinsics.rs [hunk 1]
```diff
@@ -2393,3 +2417,63 @@ where
...
     called_in_const.call_once(arg)
 }
+
+#[cfg(bootstrap)]
+pub(crate) const unsafe fn assert_validity_of<T>(_: *const T) -> bool {
+    true
+}
+
+#[cfg(not(bootstrap))]
+/// Asserts that the value at `value` is valid at type T.
```
Comment (est31) [inline library/core/src/intrinsics.rs:2427]: Is it valid to say "valid at type T"?

Comment (5225225) [reply library/core/src/intrinsics.rs:2427]: No idea, but it's what i'm used to saying. I guess "is a valid T" is a less awkward way of saying it. I'll change it to say that.

**Comment by 5225225** [comment]
Also I forgot to add them at first so I don't think highfive will ping you but hello @bjorn3 I believe rustc_codegen_cranelift is yours I'm once again messing with your intrinsics

By visual inspection it looks fine to me, I did roughly the same thing as in rustc_codegen_ssa, but is there a way for me to actually test this?

File: library/core/src/intrinsics.rs [hunk 2]
```diff
@@ -2138,6 +2138,30 @@ pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: us
...
 }
 
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct Invariant {
+    offset: u64,
+    size: u64,
+    start: u128,
+    end: u128,
+}
```
Comment (RalfJung) [inline library/core/src/intrinsics.rs:2148]: What does an `Invariant` *mean*? As in, what is the actual invariant that is encoded by an element of `Invariant`?

**Comment by bjorn3** [comment]
The cg_clif change looks fine.

File: library/core/src/intrinsics.rs (hunk 2 above)
Comment (5225225) [reply library/core/src/intrinsics.rs:2148]: Each Invariant in the list means "read off an integer with byte size `size`, `offset` bytes past the start of the struct, and then assert that it is inside the WrappingRange { start, end }. (Where WrappingRange is https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/struct.WrappingRange.html )

So for `(NonZeroU16, u8, bool)`, the list will be

* `offset: 0, size: 2, start: 1, end: 65535`
* `offset: 2, size: 1, start: 0, end: 255`
* `offset: 3, size: 1, start: 0, end: 1`.
(assuming that's laid out in order)

I'll change `size` to be an enum so the match on it is exhaustive, and so I can add "is a pointer" so we're not doing ptr to int transmutes.

We still generate an Invariant for types like `u8` so that we branch on their value, so memory sanitizer picks it up.

Though we could only enable generating invariants for types that are fully valid at any bit-pattern if memory sanitizer is enabled, since that's the only time they're useful.

Comment (RalfJung) [reply library/core/src/intrinsics.rs:2148]: Ah, I see. I think `valid_range_{start,end}` would be better names for those variables then. Please also add a doc comment explaining the meaning of the fields.

**Comment by 5225225** [comment]
This fails a test, and I'm not sure what the best way to get around it is.

It's calling mem::replace and then looking to see if any other `memcpy`s exist in the code, which this PR introduces (in debug assertions), and i don't think tests can rebuild the stdlib (without debug assertions).

I could cfg out the whole of the assertions here so you both have to build the stdlib and pass a flag to enable the assertions (like we do for `core/debug_refcell` to add locations), which might be the best bet depending on how slow they are. That does mean this code isn't being tested though.

That, or let the tests ask to rebuild the stdlib, disabling the debug assertions.

File: library/core/src/mem/maybe_uninit.rs [hunk 3]
```diff
@@ -626,11 +626,17 @@ impl<T> MaybeUninit<T> {
...
     #[rustc_diagnostic_item = "assume_init"]
     #[track_caller]
+    #[rustc_allow_const_fn_unstable(const_refs_to_cell)]
     pub const unsafe fn assume_init(self) -> T {
         // SAFETY: the caller must guarantee that `self` is initialized.
         // This also means that `self` must be a `value` variant.
         unsafe {
             intrinsics::assert_inhabited::<T>();
+
+            intrinsics::assert_unsafe_precondition!(intrinsics::assert_validity_of::<T>(
```
Comment (nagisa) [inline library/core/src/mem/maybe_uninit.rs:636]: Implementation wise, a more appropriate way would probably be to make this entire method an intrinsic/language item/whatever, rather than just the validity assertion. Doing so would also mean that the concerns about “how was libstd built” no longer matter since the callers would be calling into an intrinsic, giving an opportunity to codegen this function into the calling crate according to that crate’s flags.

File: src/test/ui/sanitize/memory.rs [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // needs-sanitizer-support
 // needs-sanitizer-memory
 //
-// compile-flags: -Z sanitizer=memory -Zsanitizer-memory-track-origins -O
+// compile-flags: -Z sanitizer=memory -Zno-validity-invariant-checks -Zsanitizer-memory-track-origins -O
```
Comment (nagisa) [inline src/test/ui/sanitize/memory.rs:4]: Before landing – the flag should be inverted and enabling it by default would need to be considered alongside stabilization of this feature as a whole.

File: src/test/ui/intrinsics/validity_invariants_of.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
+// run-pass
+// revisions: disabled normal sanitized
+// [disabled]compile-flags: -Zno-validity-invariant-checks
+// [sanitized]compile-flags: -Z sanitizer=memory
```
Comment (nagisa) [inline src/test/ui/intrinsics/validity_invariants_of.rs:4]: You will need to ignore this test based on `needs-sanitizer-*`. grep for `needs-sanitizer` in the test suite for examples of use.

File: library/core/src/lib.rs [hunk 6]
```diff
@@ -156,7 +156,6 @@
 #![feature(const_slice_from_ref)]
 #![feature(const_slice_index)]
 #![feature(const_is_char_boundary)]
-//
```
Comment (nagisa) [inline library/core/src/lib.rs:159]: Unrelated change, probably should undo (the extra space does give some sense of separation between blocks that isn’t really there with just a comment alone)

File: compiler/rustc_middle/src/mir/interpret/value.rs [hunk 7]
```diff
@@ -37,6 +37,9 @@ pub enum ConstValue<'tcx> {
     /// Used only for `&[u8]` and `&str`
     Slice { data: ConstAllocation<'tcx>, start: usize, end: usize },
 
+    /// Like `Slice`, but for types that aren't 1 byte long.
+    CustomSlice { data: ConstAllocation<'tcx>, length: usize },
```
Comment (nagisa) [inline compiler/rustc_middle/src/mir/interpret/value.rs:41]: Unclear to me why `Slice` cannot be adapted for this, but my opinion here is probably not worth much since I don’t fiddle much around `mir/interpret`.

File: src/test/ui/intrinsics/validity_invariants_of.rs (hunk 5 above)
Comment (5225225) [reply src/test/ui/intrinsics/validity_invariants_of.rs:4]: Yeah, I tried just requiring `needs-sanitizer-memory` for just the `sanitized` revision, but that doesn't seem to be allowed for a revision.

I'll split the test into two and just require the sanitizer to exist for the -Z sanitizer=memory test. Alternatively, I could try and get compiletest to allow `[sanitized]needs-sanitizer-memory`, or just put the whole test as requiring it.

**Comment by RalfJung** [comment]
> It's calling mem::replace and then looking to see if any other memcpys exist in the code, which this PR introduces (in debug assertions), and i don't think tests can rebuild the stdlib (without debug assertions).

I think there is a way to ignore tests on debug builds. `ignore-debug` or so? Some existing tests do that.

**Comment by bjorn3** [comment]
> ignore-debug or so?

Indeed

**Comment by 5225225** [comment]
Closing this since I'm unlikely to get to it any time soon, and I'll be doing it as a MIR pass instead of checks inside `core`, with not very much shared code.
